### PR TITLE
Add with_project_definition decorator

### PR DESCRIPTION
**Pull request by sfc-gh-turbaszek**
### Pre-review checklist
   * [x] I've confirmed that instructions included in README.md are still correct after my changes in the codebase.
   * [x] I've added or updated automated unit tests to verify correctness of my new code.
   * [x] I've added or updated integration tests to verify correctness of my new code.
   * [x] I've confirmed that my changes are working by executing CLI's commands manually.
   * [x] I've confirmed that my changes are up-to-date with the target branch.
   * [x] I've described my changes in the release notes.
   * [x] I've described my changes in the section below.

### Changes description
Adds `with_project_definition` decorator which unifies how command access project definition.

### CODE REVIEW

File: src/snowcli/cli/common/cli_global_context.py [hunk 1]
```diff
@@ -26,7 +26,7 @@ def __setattr__(self, key, value):
         We invalidate connection cache every time connection attributes change.
         """
         super.__setattr__(self, key, value)
-        if key is not "_cached_connection":
+        if key != "_cached_connection":
```
Comment (sfc-gh-turbaszek) [inline src/snowcli/cli/common/cli_global_context.py:29]:
```suggestion
        if key is not "_cached_connection":
```

File: src/snowcli/cli/common/cli_global_context.py [hunk 2]
```diff
@@ -184,6 +188,10 @@ def connection_context(self) -> _ConnectionContext:
     def connection(self) -> SnowflakeConnection:
         return self.connection_context.connection
 
+    @property
```
Comment (sfc-gh-pjob) [inline src/snowcli/cli/common/cli_global_context.py:191]: Can we move this property to be placed above its setter method (to keep the same convention as other properties with setters)?

File: src/snowcli/cli/common/cli_global_context.py [hunk 3]
```diff
@@ -162,6 +163,9 @@ def output_format(self) -> OutputFormat:
     def set_output_format(self, value: OutputFormat):
         self._output_format = value
 
+    def set_project_definition(self, value: Dict):
```
Comment (sfc-gh-pjob) [inline src/snowcli/cli/common/cli_global_context.py:166]: Can we move this setter (and connected property) to be placed under `set_experimental` to keep the same order as in the `init`?

File: src/snowcli/cli/common/cli_global_context.py (hunk 1 above)
Comment (sfc-gh-turbaszek) [reply src/snowcli/cli/common/cli_global_context.py]: Fixed :d

File: src/snowcli/cli/common/decorators.py [hunk 4]
```diff
@@ -49,6 +51,47 @@ def global_options_with_connection(func: Callable):
...
+
+        if not project_definition:
+            raise NoProjectDefinitionError(
+                project_type=project_name, project_file=project_path
+            )
+
+        cli_context_manager.set_project_definition(project_definition)
+        return project_definition
+
+    project_definition_option = typer.Option(
```
[the remark this answers is not in the document]
Comment (sfc-gh-turbaszek) [reply src/snowcli/cli/common/decorators.py:69]: Good point1 Moved the option and callback definition to flags.py

File: src/snowcli/cli/common/decorators.py [hunk 5]
```diff
@@ -49,6 +51,47 @@ def global_options_with_connection(func: Callable):
     )
 
 
+def with_project_definition(project_name: str):
+    from snowcli.cli.project.definition_manager import DefinitionManager
+
+    def _callback(project_path: str):
```
[the remark this answers is not in the document]
Comment (sfc-gh-turbaszek) [reply src/snowcli/cli/common/decorators.py:57]: Fixed

**Review by sfc-gh-pjob — APPROVED** [review]

File: src/snowcli/cli/common/cli_global_context.py [hunk 6]
```diff
@@ -219,6 +227,10 @@ def verbose(self) -> bool:
     def experimental(self) -> bool:
         return self._manager.experimental
 
+    @property
+    def project_definition(self):
+        return self._manager.project_definition
```
Comment (sfc-gh-astus) [inline src/snowcli/cli/common/cli_global_context.py:232]: Is it ok to have `project_definition` property twice?

Comment (sfc-gh-pjob) [reply src/snowcli/cli/common/cli_global_context.py:232]: It is in a different class :)

**Review by sfc-gh-pjob — APPROVED** [review]

**Pull request merged**